### PR TITLE
fix incorrect image size when rendition is inserted 

### LINKED ISSUE #1806: Image uploader displays incorrect image size when rendition is inserted
### Steps to reproduce (*)
Insert a large image from media gallery to image-uploader (i.e. category image)

### Expected result (*)
Rendition image size should be displayed

### Actual result (*)
image

### PR DESCRIPTION
**Pull request by engcom-Golf**
### Description (*)

Fixes issue that image uploader displays incorrect image size when rendition is inserted from media gallery
### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1806

#### Preconditions:
1. Use old media gallery - No
2. Media Gallery Renditions Configured
#### Steps to test:
1. Open category page
2. Open content section
3. Insert Image from media gallery
4. Insert image with high resolution and file size
5. Ensure that image resolution and size responds to rendition image

### CODE REVIEW

File: app/code/Magento/Cms/Model/Wysiwyg/Images/GetInsertImageContent.php [hunk 1]
```diff
@@ -60,4 +88,43 @@ public function execute(
...
         return $this->imagesHelper->getImageHtmlDeclaration($filename, $renderAsTag);
     }
+
+    /**
+     * Retrieve size of requested file
+     *
+     * @param string $path
+     * @return int
+     */
+    public function getImageSize(string $path): int
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/Wysiwyg/Images/GetInsertImageContent.php]: Please move these public methods to the introduced controller (as private methods).

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @engcom-Golf ! Please see my comment.

Also can you please add an MFTF test covering the fix

File: app/code/Magento/MediaGalleryUi/Ui/Component/Listing/Columns/Url.php [hunk 2]
```diff
@@ -98,7 +98,7 @@ public function prepare(): void
             array_replace_recursive(
                 (array)$this->getData('config'),
                 [
-                    'onInsertUrl' => $this->urlInterface->getUrl('cms/wysiwyg_images/oninsert'),
+                    'onInsertUrl' => $this->urlInterface->getUrl('media_gallery_renditions/image/oninsert'),
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Ui/Component/Listing/Columns/Url.php]: The issue with this approach is that it makes MediaGalleryUi module dependent on Renditions. It's better to have the reversed dependency.

Let's keep the plugin in Renditions module and move the introduced controller to MediaGalleryUi for this purpose

**Comment by engcom-Golf** [comment]
@magento run all tests

File: app/code/Magento/Cms/Model/Wysiwyg/Images/GetInsertImageContent.php [hunk 3]
```diff
@@ -10,6 +10,11 @@
 
 use Magento\Catalog\Helper\Data as CatalogHelper;
 use Magento\Cms\Helper\Wysiwyg\Images as ImagesHelper;
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\File\Mime;
+use Magento\Framework\Filesystem;
+use Magento\Framework\Filesystem\Directory\ReadInterface;
+use Magento\Framework\Filesystem\File\WriteInterface;
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/Wysiwyg/Images/GetInsertImageContent.php]: Please remove the unused imports

File: app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminMediaGalleryInsertLargeImageFileSizeTest">
+        <annotations>
+            <features value="AdminMediaGalleryInsertLargeImageFileSizeTest"/>
+            <useCaseId value="https://github.com/magento/adobe-stock-integration/issues/1806"/>
+            <title value="Admin user should see correct image file size after rendition"/>
+            <testCaseId value="https://studio.cucumber.io/projects/131313/test-plan/folders/1507933/scenarios/5200023"/>
+            <description value="Admin user should see correct image file size after rendition"/>
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml:18]: Please add stories tag, value: "User inserts image rendition to the content"

File: app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            <severity value="MAJOR"/>
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml]:
```suggestion
            <severity value="AVERAGE"/>
```

File: app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            <!-- Delete category -->
+            <deleteData createDataKey="category" stepKey="deleteCategory"/>
+        </after>
+
+        <!-- Open category page -->
+        <actionGroup ref="AdminOpenCategoryGridPageActionGroup" stepKey="openCategoryPage"/>
+        <actionGroup ref="AdminEditCategoryInGridPageActionGroup" stepKey="editCategoryItem">
+            <argument name="categoryName" value="$category.name$"/>
+        </actionGroup>
+        <actionGroup ref="AdminSaveCategoryFormActionGroup" stepKey="saveCategoryForm"/>
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryRenditions/Test/Mftf/Test/AdminMediaGalleryInsertLargeImageFileSizeTest.xml]: It looks like an unnecessary step

File: app/code/Magento/MediaGalleryUi/Test/Mftf/Data/AdminEnhancedMediaGalleryImageData.xml [hunk 7]
```diff
@@ -40,4 +40,13 @@
         <data key="extension">jpg</data>
         <data key="keywords">magento, mediagallerymetadata</data>
     </entity>
+    <entity name="LargeImage" type="image">
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Test/Mftf/Data/AdminEnhancedMediaGalleryImageData.xml]: I don't think new entity is needed. `ImageUpload` can be used (it's referencing the same file)

File: app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Return a content (just a link or an html block) for inserting image to the content
+     *
+     * @return ResultInterface
+     */
+    public function execute()
+    {
+        $data = $this->getRequest()->getParams();
+        $path = $this->getInsertImageContent->execute(
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php]: That is actually a content (that is some cases may be just a path). Please correct naming on backend and frontent
```suggestion
        $content = $this->getInsertImageContent->execute(
```

File: app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            $data['filename'],
+            $data['force_static_path'],
+            $data['as_is'],
+            isset($data['store']) ? (int)$data['store'] : null
+        );
+
+        $size = $data['force_static_path'] ? $this->getImageSize($path) : 0;
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php:87]: I think it's better to use `$data['filename']` for retrieving the file path for getting more information

File: app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        return $this->resultJsonFactory->create()->setData(['path' => $path, 'size' => $size, 'type' => $type]);
+    }
+
+    /**
+     * Retrieve size of requested file
+     *
+     * @param string $path
+     * @return int
+     */
+    private function getImageSize(string $path): int
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php]: considering the amount of business logic, I'd extract it to a `Magento\MediaGalleryUi\Model\GetInsertData` model.

I'd have a public method in this model with the following interface:
public function execute(
        string $encodedFilename,
        bool $forceStaticPath,
        bool $renderAsTag,
        ?int $storeId = null
    ): \Magento\MediaGalleryUi\Model\InsertDataInterface

`InsertDataInterface` should be an extensible interface (extending `\Magento\Framework\Api\ExtensibleDataInterface`) implemented by `\Magento\MediaGalleryUi\Model\InsertData` that should handle extension attributes

`InsertDataInterface` should have methods: getPath, getSize, getType,setExtensionAttributes,getExtensionAttributes

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for updates @engcom-Golf ! Please see my comments

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

File: app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryUiApi\Api;
+
+use Magento\Framework\Api\ExtensibleDataInterface;
+
+/**
+ * Class responsible to provide insert image details
+ */
+interface GetInsertImageDataInterface extends ExtensibleDataInterface
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php]:
```suggestion
interface InsertAssetDataInterface extends ExtensibleDataInterface
```

File: app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Retrieves a content (just a link or an html block) for inserting image to the content
+     *
+     * @param string $encodedFilename
+     * @param bool $forceStaticPath
+     * @param bool $renderAsTag
+     * @param int|null $storeId
+     * @return null|string
+     */
+    public function getImageContent(
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php]: This method should be part of the service, it shouldn't be present in data transfer object https://en.wikipedia.org/wiki/Data_transfer_object

File: app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        $data = $this->getRequest()->getParams();
+        $content = $this->getInsertImageData->getImageContent(
+            $data['filename'],
+            $data['force_static_path'],
+            $data['as_is'],
+            isset($data['store']) ? (int)$data['store'] : null
+        );
+
+        $size = $data['force_static_path'] ? $this->getInsertImageData->getFileSize($content) : 0;
+        $type = $data['force_static_path'] ? $this->getInsertImageData->getMimeType($content) : '';
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php]: Size and type retrieving logic should be moved to `getInsertImageData` service.

The service should retrieve all required data and create an `InsertImageData` (`insertImageDataFactory->create(['data' => ['content' => ..., 'size' => ..., ...]])`) and return the `InsertImageData` to the conroller.

File: app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        return $this->resultJsonFactory->create()->setData(['content' => $content, 'size' => $size, 'type' => $type]);
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Controller/Adminhtml/Image/OnInsert.php]: The controller should retrieve all the infromation from the data transfer objectЖ
```suggestion
        return $this->resultJsonFactory->create()->setData(
        [
            'content' => $insertImageData->getContent(), 
            'size' => $insertImageData->getSize(), 
            'type' => $insertImageData->getType()
        ]
        );
```

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @engcom-Golf , thanks for updates! Please see my comments

**Comment by engcom-Golf** [comment]
@magento run Static Tests, Functional Tests CE

File: app/code/Magento/MediaGalleryUi/Model/InsertImageData.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        return $this->extensionAttributes;
+    }
+
+    /**
+     * Set extension attributes
+     *
+     * @param ?\Magento\MediaGalleryUiApi\Api\InsertImageDataExtensionInterface;|null $extensionAttributes
+     * @return void
+     */
+    public function setExtensionAttributes(?InsertImageDataExtensionInterface $extensionAttributes): void
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Model/InsertImageData.php]: I don't see the use case when we should allow calling `setExtensionAttributes` passing `null` argument
```suggestion
    public function setExtensionAttributes(InsertImageDataExtensionInterface $extensionAttributes): void
```

File: app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryUiApi\Api;
+
+use Magento\MediaGalleryUiApi\Api\Data\InsertImageDataInterface;
+
+/**
+ * Class responsible to provide insert image details
+ */
+interface GetInsertImageDataInterface
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUiApi/Api/GetInsertImageDataInterface.php]: I don't think this interface is necessary

File: app/code/Magento/MediaGalleryUiApi/Api/Data/InsertImageDataInterface.php [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    public function getExtensionAttributes(): ?\Magento\MediaGalleryUiApi\Api\InsertImageDataExtensionInterface;
+
+    /**
+     * Set extension attributes
+     *
+     * @param \Magento\MediaGalleryUiApi\Api\InsertImageDataExtensionInterface|null $extensionAttributes
+     * @return void
+     */
+    public function setExtensionAttributes(
+        ?\Magento\MediaGalleryUiApi\Api\InsertImageDataExtensionInterface $extensionAttributes
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUiApi/Api/Data/InsertImageDataInterface.php]:
```suggestion
        \Magento\MediaGalleryUiApi\Api\InsertImageDataExtensionInterface $extensionAttributes
```

File: app/code/Magento/MediaGalleryUiApi/Api/Data/InsertImageDataInterface.php [hunk 18]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryUiApi\Api\Data;
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUiApi/Api/Data/InsertImageDataInterface.php]: Let's move this interface to the suggested namespace
```suggestion
namespace Magento\MediaGalleryUi\Model;
```

File: app/code/Magento/MediaGalleryUi/Model/InsertImageData/GetInsertImageData.php [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+     * Retrieve MIME type of requested file
+     *
+     * @param string $path
+     * @return string
+     */
+    public function getType(string $path): string
+    {
+        $absoluteFilePath = $this->getPubDirectory()->getAbsolutePath($path);
+
+        return $this->getPubDirectory()->isExist($path) ? $this->mime->getMimeType($absoluteFilePath) : '';
```
Comment (sivaschenko) [inline app/code/Magento/MediaGalleryUi/Model/InsertImageData/GetInsertImageData.php]: I'd check if path exists before even building absolute path
```suggestion
        return $this->getPubDirectory()->isExist($path) ? $this->mime->getMimeType($this->getPubDirectory()->getAbsolutePath($path)) : '';
```

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Perfect @engcom-Golf ! Probably the final comments from me are regarding interfaces:

I this the data interface should be moved to the implementation module and the service interface is not needed currently.

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run Functional Tests CE

**Comment by engcom-Golf** [comment]
@magento give me test instance

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-8217](https://jira.corp.magento.com/browse/ENGCOM-8217) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

Correct image size is displayed before saving the Category

See the result **before** switching to the PR
before29979

and the result **after** switching to the PR 
after29979

Manual testing scenario - https://studio.cucumber.io/projects/131313/test-plan/folders/1507933/scenarios/5200023

**Pull request merged**